R/module_snapshot_manager.R

```diff
@@ -0,0 +1,276 @@
+        )
+        updateTextInput(inputId = "snapshot_name", value = , placeholder = "Meaningful, unique name")
+      } else {
+        snapshot <- disassemble_slices(slices_global())
+        attr(snapshot, "mapping") <- matrix_to_mapping(mapping_matrix())
+        snapshot_update <- c(snapshot_history(), list(snapshot))
+        names(snapshot_update)[length(snapshot_update)] <- snapshot_name
+        snapshot_history(snapshot_update)
+        removeModal()
+        # Reopen filter manager modal by clicking button in the main application.
+        shinyjs::click(id = "teal-main_ui-filter_manager-show", asis = TRUE)
+      }
+    })
+
+    # Restore initial state.
+    observeEvent(input$snapshot_reset, {
+      s <- "Initial application state"
+      ### Begin restore procedure. ###
+      snapshot <- snapshot_history()[[s]]
+      snapshot_state <- reassemble_slices(snapshot)
+      mapping_unfolded <- unfold_mapping(attr(snapshot_state, "mapping"), names(filtered_data_list))
+      mapply(
+        function(filtered_data, filters) {
+          filtered_data$clear_filter_states(force = TRUE)
+          slices <- Filter(function(x) x$id %in% filters, snapshot_state)
+          filtered_data$set_filter_state(slices)
+        },
+        filtered_data = filtered_data_list,
+        filters = mapping_unfolded
+      )
+      slices_global(snapshot_state)
+      removeModal()
+      ### End restore procedure. ###
+    })
+
+    # Create table to display list of snapshots and their actions.
+    output$snapshot_list <- renderUI({
+      lapply(names(snapshot_history())[-1L], function(s) {
+        id_pickme <- sprintf("pickme_%s", make.names(s))
+        id_saveme <- sprintf("saveme_%s", make.names(s))
+
+        # Restore snapshot.
+        observeEvent(input[[id_pickme]], {
+          ### Begin restore procedure. ###
+          snapshot <- snapshot_history()[[s]]
+          snapshot_state <- reassemble_slices(snapshot)
+          mapping_unfolded <- unfold_mapping(attr(snapshot_state, "mapping"), names(filtered_data_list))
+          mapply(
+            function(filtered_data, filters) {
+              filtered_data$clear_filter_states(force = TRUE)
+              slices <- Filter(function(x) x$id %in% filters, snapshot_state)
+              filtered_data$set_filter_state(slices)
+            },
+            filtered_data = filtered_data_list,
+            filters = mapping_unfolded
+          )
+          slices_global(snapshot_state)
+          removeModal()
+          ### End restore procedure. ###
+        })
+
+        # Save snapshot.
+        output[[id_saveme]] <- downloadHandler(
+          filename = function() {
+            sprintf("teal_snapshot_%s_%s.json", s, Sys.Date())
+          },
+          content = function(file) {
+            snapshot <- snapshot_history()[[s]]
+            snapshot_state <- reassemble_slices(snapshot)
+            teal.slice::slices_store(tss = snapshot_state, file = file)
+          }
+        )
```

Comment:
observers are reinstantiated for all snapshots each time when `snapshot_history()` changes. `observeEvent` and `downloadHandler` should be outside of `renderUI` and instantiated only once per snapshot.

Reply:
They do not seem to be duplicated though. No matter how many times snapshot history is updated, a click on a save button only triggers one handler, a click on a select button restores the state once (no visible multiple updates to the filter panel). Are you sure this is a problem?

I have rewritten it (not pushed yet) so that observers, handlers and divs are each stored in separate `reactiveValues` and one observer on `snapshot_history` only appends elements that are not already there but does not recreate the whole list. This way observers and handlers are instantiated only once. Would that be better?

Reply:
Yes, click will trigger one handler. And yes, they are not duplicated, duplicated is operation of setting these observers.

> Are you sure this is a problem?

Doesn't seem right to create new observer in a session and overwrite the previous one. I'm pretty sure there some side effects in a `session` object because of this (relinking perhaps).

> Would that be better?

Hard to imagine final look. 

-----

What if div stays in `renderUI` and observers are moved to `observeEvent(added_snapshots(), ...`, similarly to this, see lines 363-394:

https://github.com/insightsengineering/teal.slice/blob/3f4c6ff49eb9f7816fed7849af53bfd431e85484/R/FilterStates.R#L375

Reply:
I thought about smth similar but I did not want to juggle multiple `reactiveVal`s as I have 2 or 3 lists to maintain here. Let me push what I described above (the current version will be commented out) so you can have a look. It is not much different.

Reply:
I had my shiny lesson now. Invalidation between elements happens from one-to-another and session is not involved in the process. I was wrong in the first comment as adding new observer doesn't update whole reactivity graph - it is smart enough to pass invalidation locally between two elements (reactive -> endpoint)

Reply:
@chlebowa you can back to the state before my first comment. This scenario differs from `srv_active` as we only add snapshots. so we don't need to remove observers at any point.

P.S. Update "do not reinstantiate observers" hasn't changed anything from this perspective as there is `observeEvent(snapshot_history(), ... observeEvent() ...)`

Reply:
After some talking we decided to create UI elements and server logic for individual snapshots outside of `renderUI`. They will be stored in `reactiveValues` and new ones will be appended only if they do not yet exist.